docs/contributors/design-proposals/mutable-operator-registry.md

```diff
@@ -0,0 +1,125 @@
+
+## Proposal
+
+This implementation makes the assumption that a separate process is built that generates container images that contain individual operator bundles for each version of an operator.
+
+To start, let's define what the operator-registry does today. It:
+
+1. Takes a set of manifest files and outputs a database file.
+2. Takes a database file and serves a grpc api that serves content from that db
+3. Takes a configmap reference, builds a db and serves a grpc api that serves content from that db
+4. Takes an app-registry path/namespace reference, builds a db and serves a grpc api that serves content from that db
+
+And what we would like it to:
+
+1. Use a reference to a container image (also referred to here as the operator index) rather than app-registry namespaces to expose installable operators on a cluster
+2. Build an operator-registry database that serves the data required by OLM to drive install and UI workflows
+3. Have a way to optimize the build workflow of operator-registry databases (which currently drive OLM's workflows) from previous versions of the database plus new content (ex. new operator at new version) so that the database need not be built from scratch every time they are created.
+
+### Updating the Operator Registry to insert incremental updates
+
+We want to add create db, delete, and batch insert APIs to the model layer of operator-registry and a new set of operator registry commands to utilize those new APIs:
+
+`operator-registry create`
+
+- inputs: none
+
+- outputs: empty operator registry database
+    
+`operator-registry add`
+
+- inputs: $operatorBundleImagePath, $dbFile
+
+- outputs: updated database file
+
+ex: `operator-registry add quay.io/community-operators/foo:sha256@abcd123 example.db`
+   
+`operator-registry add --batch`
+
+- inputs: $operatorBundleImagePath, $dbFile
+
+- outputs: updated database file
+
+ex: `operator-registry add "quay.io/community-operators/foo:sha256@abcd123,quay.io/community-operators/bar:sha256@defg456" example.db`
+    
+`operator-registry delete`
+
+- inputs: $operatorName, $dbFile
+
+- outputs: updated database file without $operatorName included
+
+ex: `operator-registry delete bar example.db`
+
+`operator-registry delete-latest`
+
+- inputs: $operatorName, $dbFile
+
+- outputs: updated database file without latest version of $operatorName included
+
+ex: `operator-registry delete-latest foo example.db`
+
+### Implementation details
+
+#### Add
+
+To achieve this we need to pull down images from a registry using podman, extract the bundle files from the container's filesystem and insert the bundle into the database. For each channel defined in the bundle's `package.yaml`:
```

Comment:
I'm wondering if we want to have a common user experience to the `operator-sdk` in this case. I do know that the sdk uses environment variables and wraps their build commands (for example) but under the hood they are using podman/buildah/docker cli: https://github.com/operator-framework/operator-sdk/blob/master/cmd/operator-sdk/build/cmd.go#L62

If the sdk is going to call these commands as functions, should we attempt to also have the same experience there? We could even just default to containerd if we wanted to, but from the perspective of a user trying to wrangle all these tools together does it make more sense that they all have the same defaults?

Reply:
This reminds me a lot of the [Interface Segregation Principle](https://en.wikipedia.org/wiki/Interface_segregation_principle). We don't require everything that SDK does, so why take on the burdens of requiring podman? I don't think the UX of this tool will differ since we're already abstracting over the operation.